src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp

```diff
@@ -0,0 +1,758 @@
+   BOTAN_ASSERT_NOMSG(rhoprime.size() == DilithiumModeConstants::CRHBYTES);
+   BOTAN_ASSERT_NOMSG(key.size() == DilithiumModeConstants::SEEDBYTES);
+
+   /* Generate matrix */
+   auto matrix = Dilithium::PolynomialMatrix::generate_matrix(rho, mode);
+
+   /* Sample short vectors s1 and s2 */
+   Dilithium::PolynomialVector s1(mode.l());
+   Dilithium::PolynomialVector::fill_polyvec_uniform_eta(s1, rhoprime, 0, mode);
+
+   Dilithium::PolynomialVector s2(mode.k());
+   Dilithium::PolynomialVector::fill_polyvec_uniform_eta(s2, rhoprime, mode.l(), mode);
+
+   auto [ t0, t1 ] = calculate_t0_and_t1(mode, rho, s1, s2);
+
+   m_public = std::make_shared<Dilithium_PublicKeyInternal>(mode, rho, std::move(t1));
+
+   /* Compute H(rho, t1) == H(pk) and write secret key */
+   const auto tr = mode.H(m_public->raw_pk(), DilithiumModeConstants::SEEDBYTES);
+
+   m_private = std::make_shared<Dilithium_PrivateKeyInternal>(std::move(mode), std::move(rho), std::move(tr), std::move(key),
+                                                              std::move(s1), std::move(s2), std::move(t0));
+   }
+
+Dilithium_PrivateKey::Dilithium_PrivateKey(const AlgorithmIdentifier& alg_id, const secure_vector<uint8_t>& sk)
+   : Dilithium_PrivateKey(sk, dimension_from_alg_id(alg_id), algo_from_alg_id(alg_id), DilithiumKeyEncoding::Raw) {}
+
+Dilithium_PrivateKey::Dilithium_PrivateKey(const secure_vector<uint8_t>& sk,
+                                           DilithiumDimension dimension, DilithiumFlavor algo, DilithiumKeyEncoding encoding)
+   {
+   DilithiumModeConstants mode(dimension, algo);
+
+   switch(encoding)
+      {
+      case Botan::DilithiumKeyEncoding::Raw:
+         BOTAN_ARG_CHECK(sk.size() == mode.private_key_bytes(),
+                         "dilithium private key does not have the correct byte count");
+         m_private = std::make_shared<Dilithium_PrivateKeyInternal>(std::move(mode), sk);
+         break;
+      case Botan::DilithiumKeyEncoding::DER:
+         m_private = std::make_shared<Dilithium_PrivateKeyInternal>(std::move(mode));
+         BER_Decoder(sk).decode(*m_private);
+         break;
+      }
+
+   m_public = std::make_shared<Dilithium_PublicKeyInternal>(m_private->mode(), m_private->rho(), m_private->s1(), m_private->s2());
+   }
+
+secure_vector<uint8_t> Dilithium_PrivateKey::private_key_bits() const
+   {
+   switch(m_key_encoding)
+      {
+      case DilithiumKeyEncoding::Raw:
+         return m_private->raw_sk();
+
+      case DilithiumKeyEncoding::DER:
+         secure_vector<uint8_t> out;
+         DER_Encoder(out).encode(*m_private);
+         return out;
+      }
+
+   unreachable();
+   }
+
+std::unique_ptr<PK_Ops::Signature>
+Dilithium_PrivateKey::create_signature_op(RandomNumberGenerator& rng,
+      const std::string& params,
+      const std::string& provider) const
+   {
+   BOTAN_UNUSED(rng);
+   BOTAN_UNUSED(params);
+   BOTAN_UNUSED(provider);
```

Comment:
nit: you can use `BOTAN_UNUSED(rng, params, provider);`

Reply:
Actually that is not really right here - to allow future extension, we should check that params is empty and that provider is either empty or "base"

Reply:
Changed to:

```C++
   BOTAN_UNUSED(rng);
   BOTAN_ARG_CHECK(params.empty(), "Unexpected parameters for signing with Dilithium");
   if(provider.empty() || provider == "base")
      return std::make_unique<PK_Ops::Signature>(*this);
   throw Provider_Not_Found(algo_name(), provider);
```

I reckon that `params` might be used for the Deterministic/Randomized refactoring later.

Reply:
A similar change was applied to `create_verification_op()`.